sklearn/preprocessing/tests/test_discretization.py

```diff
@@ -94,6 +145,21 @@ def test_fit_transform_n_bins_array(strategy, expected):
         assert bin_edges.shape == (n_bins + 1,)
 
 
+def test_zero_sample_weight():
+    X = np.array([[-2], [-1], [1], [3], [500], [1000]])
+    est = KBinsDiscretizer(n_bins=2, encode="ordinal", strategy="quantile")
+    est.fit(X, sample_weight=[1, 1, 1, 1, 0, 0])
+    assert_array_equal([[0.0], [1.0], [1.0], [1.0], [1.0], [1.0]], est.transform(X))
```

Comment:
I think that we can slightly improve the test such that we can as well check the bin edges generated

```suggestion
def test_kbinsdiscretizer_effect_sample_weight():
    """Check that we take into account `sample_weight` when computing the quantiles."""
    X = np.array([[-2], [-1], [1], [3], [500], [1000]])
    # add a large number of bins such that each sample with a non-null weight
    # will be used as bin edge
    est = KBinsDiscretizer(n_bins=10, encode="ordinal", strategy="quantile")
    est.fit(X, sample_weight=[1, 1, 1, 1, 0, 0])
    assert_allclose(est.bin_edges_[0], [-2, -1, 1, 3])
    assert_allclose(est.transform(X), [[0.0], [1.0], [2.0], [2.0], [2.0], [2.0]])
```

Be aware that you can add the import for the testing as:

```python
from sklearn.utils._testing import assert_allclose
```